### ISSUE TITLE
Incompatible UPDATE with MySQL

### ISSUE THREAD

**Issue by sayJason** [pingcap/tidb#32857]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
DROP TABLE IF EXISTS t0; 
CREATE TABLE t0 (c0 DECIMAL, c1 TEXT); 
INSERT INTO t0 VALUES (1, 'a'); 
UPDATE t0 SET c0 = 2 WHERE t0.c0 > t0.c1;
```

### 2. What did you expect to see? (Required)
TiDB should succeed to update the value like MySQL
### 3. What did you see instead (Required)
```sql
mysql> UPDATE t0 SET c0 = 2 WHERE t0.c0 > t0.c1;
ERROR 1292 (22007): Truncated incorrect INTEGER value: 'a'
```
### 4. What is your TiDB version? (Required)
```sql
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()

                                                                       |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.4.0
Edition: Community
Git Commit Hash: 55f3b24c1c9f506bd652ef1d162283541e428872
Git Branch: heads/refs/tags/v5.4.0
UTC Build Time: 2022-01-25 08:39:26
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by sayJason** [comment]
I notice there is a similar case in issue https://github.com/pingcap/tidb/issues/31405, where the following UPDATE statement succeeded.
```sql
DROP TABLE IF EXISTS t0; 
CREATE TABLE t0 (c0 DECIMAL); 
INSERT INTO t0 VALUES (1); 
UPDATE t0 SET c0 = 2 WHERE t0.c0 > 'a';
```
Is there any difference to evaluate the expression between 'a' is a constant and 'a' is the column value?  I think there should be no difference.

**Comment by fanrenhoo** [comment]
> I notice there is a similar case in issue #31405, where the following UPDATE statement succeeded.
> 
> ```sql
> DROP TABLE IF EXISTS t0; 
> CREATE TABLE t0 (c0 DECIMAL); 
> INSERT INTO t0 VALUES (1); 
> UPDATE t0 SET c0 = 2 WHERE t0.c0 > 'a';
> ```
> 
> Is there any difference to evaluate the expression between 'a' is a constant and 'a' is the column value? I think there should be no difference.

I believe this should be executed by coprocessor in TiKV side, so not strange because TiDB is not MySQL just use its syntax.

**Comment by sayJason** [comment]
> > I notice there is a similar case in issue #31405, where the following UPDATE statement succeeded.
> > ```sql
> > DROP TABLE IF EXISTS t0; 
> > CREATE TABLE t0 (c0 DECIMAL); 
> > INSERT INTO t0 VALUES (1); 
> > UPDATE t0 SET c0 = 2 WHERE t0.c0 > 'a';
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is there any difference to evaluate the expression between 'a' is a constant and 'a' is the column value? I think there should be no difference.
> 
> I believe this should be executed by coprocessor in TiKV side, so not strange because TiDB is not MySQL just use its syntax.

I think the same type conversion happens in current case and case https://github.com/pingcap/tidb/issues/31405, but their results are different. The current case failed, which the other case succeeded. I feel confused and think it's an inconsistent behavior in TiDB.
Since TiDB is a weakly typed system,  I wonder are there some rules that define exactly which type conversions are allowed or not?

**Comment by fanrenhoo** [comment]
> > > I notice there is a similar case in issue #31405, where the following UPDATE statement succeeded.
> > > ```sql
> > > DROP TABLE IF EXISTS t0; 
> > > CREATE TABLE t0 (c0 DECIMAL); 
> > > INSERT INTO t0 VALUES (1); 
> > > UPDATE t0 SET c0 = 2 WHERE t0.c0 > 'a';
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Is there any difference to evaluate the expression between 'a' is a constant and 'a' is the column value? I think there should be no difference.
> > 
> > 
> > I believe this should be executed by coprocessor in TiKV side, so not strange because TiDB is not MySQL just use its syntax.
> 
> I think the same type conversion happens in current case and case #31405, but their results are different. The current case failed, which the other case succeeded. I feel confused and think it's an inconsistent behavior in TiDB. Since TiDB is a weakly typed system, I wonder are there some rules that define exactly which type conversions are allowed or not?

Yes, from this point of view, this is a problem. but may need to add type convert logic in TIKV coprocessor side either. I'm just saying its understandable for TIDB have this problem. TIDB comply with Mysql, but TIKV can be a standalone product so not sure if should take the same logic exactly same as MySQL, not sure if this fix whether or not affect other usage of TIKV
